src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs

```diff
@@ -580,8 +580,12 @@ private void FeedXRefMap(List<ManifestItemWithContext> manifest, IDocumentBuildC
                 using (new LoggerFileScope(m.FileModel.LocalPathFromRepoRoot))
                 {
                     Logger.LogDiagnostic($"Feed xref map from template for {m.Item.DocumentType}...");
-                    var bookmarks = m.Options.Bookmarks;
-                    // TODO: Add bookmarks to xref
+                    // TODO: use m.Options.Bookmarks directly after all templates report bookmarks
+                    var bookmarks = m.Options.Bookmarks ?? m.FileModel.Bookmarks;
```

Comment:
Template needn't collect, but decide bookmark info, then report it to DocFX for solving xref, and fill into mustache.

---
In reply to: [86513579](https://github.com/dotnet/docfx/pull/860#discussion_r86513579) [](ancestors = 86513579)